From ec6f241bc1c5425bc2a8f5f20fe1b8f8c2dcbd38 Mon Sep 17 00:00:00 2001
From: arathinx <ayyappanx.rathinam@intel.com>
Date: Wed, 16 Oct 2019 13:46:55 +0800
Subject: [PATCH] PONRTSYS-5441: bootrom fails to boot after reboot due to
 kernel panic

Basically, there is a known hardware issue on PRX300 global sw_reset (GLB_SW), Setting this bit(30) will not reset GPHY module. This eventually causes issue hangup with reboot command. Added explicit GPHY reset inside RCU driver itself to resolve the issue.
---
 drivers/reset/reset-intel-syscon.c | 34 ++++++++++++++++++++++++++--------
 1 file changed, 26 insertions(+), 8 deletions(-)

diff --git a/drivers/reset/reset-intel-syscon.c b/drivers/reset/reset-intel-syscon.c
index 66263cef8560..a88253b80908 100644
--- a/drivers/reset/reset-intel-syscon.c
+++ b/drivers/reset/reset-intel-syscon.c
@@ -29,7 +29,8 @@ struct intel_reset_data {
 	struct device *dev;
 	struct regmap *regmap;
 	const struct intel_reset_soc_data *soc_data;
-	u32 reboot_id;
+	u32 *rst_ids;
+	u32 rst_num;
 };
 
 static u32 intel_stat_reg_off(struct intel_reset_data *data, u32 req_off)
@@ -142,22 +143,28 @@ static int intel_reset_xlate(struct reset_controller_dev *rcdev,
 static int intel_reset_restart_handler(struct notifier_block *nb,
 				       unsigned long action, void *data)
 {
+	int cnt;
 	struct intel_reset_data *reset_data =
 		container_of(nb, struct intel_reset_data, restart_nb);
 
-	intel_assert_device(&reset_data->rcdev, reset_data->reboot_id);
+	for (cnt = 0; cnt < reset_data->rst_num; cnt++) {
+		if (reset_data->rst_ids[cnt]) {
+			intel_assert_device(&reset_data->rcdev,
+					    reset_data->rst_ids[cnt]);
+		}
+	}
 
 	return NOTIFY_DONE;
 }
 
 static int intel_reset_probe(struct platform_device *pdev)
 {
-	int err;
+	int err, cnt;
 	struct device_node *np = pdev->dev.of_node;
 	struct device *dev = &pdev->dev;
 	struct intel_reset_data *data;
 	struct regmap *regmap;
-	u32 rb_id[2];
+	struct of_phandle_args rst;
 
 	data = devm_kzalloc(dev, sizeof(*data), GFP_KERNEL);
 	if (!data)
@@ -175,14 +182,25 @@ static int intel_reset_probe(struct platform_device *pdev)
 		return -ENODEV;
 	}
 
-	if (device_property_read_u32_array(dev,
-					   "intel,global-reset", rb_id, 2)) {
-		dev_err(dev, "Failed to get global reset offset!\n");
+	data->rst_num = of_count_phandle_with_args(np,  "intel,global-reset",
+						   "#reset-cells");
+	if (data->rst_num <= 0) {
+		dev_err(dev, "intel,global-reset parsing err\n");
 		return -EINVAL;
 	}
 
+	data->rst_ids = devm_kcalloc(dev, data->rst_num, sizeof(*data->rst_ids),
+				     GFP_KERNEL);
+	if (!data->rst_ids)
+		return -ENOMEM;
+
+	for (cnt = 0; cnt < data->rst_num; cnt++) {
+		if (!of_parse_phandle_with_fixed_args(np, "intel,global-reset",
+						      2, cnt, &rst))
+			data->rst_ids[cnt] = ((rst.args[0] << 8) | rst.args[1]);
+	}
+
 	data->dev = dev;
-	data->reboot_id = (rb_id[0] << 8) | rb_id[1];
 	data->regmap = regmap;
 	data->rcdev.of_node = np;
 	data->rcdev.owner = dev->driver->owner;
